File: Using_Java/LinkedLists/DoublyLinearLL.java
class NodeD
{
	public int iData;
	public NodeD nextPtr;
	public NodeD prevPtr;
	
	public NodeD()
	{
		this.iData 	= 0;
		this.nextPtr 	= null;
		this.prevPtr  	= null;	
	}
	
	public NodeD(int iValue)
	{
		this.iData   	= iValue;
		this.nextPtr 	= null;
		this.prevPtr	= null;
	}
	
}

interface LinkedList
{
	public void insertFirst(int iNo);
	public void insertLast(int iNo);
	
	public void deleteFirst();
	public void deleteLast();
	
	public void insertAtPos(int iNo,int iPos);
	public void deleteAtPos(int iPos);
}

class DoublyLinear implements LinkedList 
{
	public NodeD first;
	public int iCount;
				
	public DoublyLinear()
	{
		this.first 	= null;
		this.iCount	= 0;
	}
	
	public void displayNodes()
	{
		int iCnt = 0;
		
		NodeD tempPtr = first;
		
		for(iCnt = 0; iCnt < iCount; iCnt++)
		{
			System.out.print("|"+ tempPtr.iData+"|<=>");
			tempPtr = tempPtr.nextPtr;
		}
		System.out.println("null");	
	}
	
	public void displayNodesInReverse()
	{
		int iCnt = 0;
		
		NodeD tempPtr = first;
	
		for(iCnt = 0; iCnt < iCount; iCnt++)
		{
			System.out.print("|"+ tempPtr.iData+"|<=>");
			tempPtr = tempPtr.prevPtr;
		}
		System.out.println("null");	
	}
	
	public int countNodes()
	{
		return this.iCount;
	}				
	
	public void insertFirst(int iNo)
	{
		NodeD newNode = null;
		
		newNode = new NodeD(iNo);
	
		if(first == null)
		{
			first = newNode;
		}
		else
		{
			newNode.nextPtr = first;
			first.prevPtr = newNode;
			first = newNode;
		}
		iCount++;
	}
	
	public void insertLast(int iNo)
	{
		NodeD newNode = null;
		
		newNode = new NodeD(iNo);
		
		if(first == null)
		{
			first = newNode;
		}
		else
		{
			NodeD tempPtr = first;
			
			while(tempPtr.nextPtr != null)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			newNode.prevPtr = tempPtr;
			tempPtr.nextPtr = newNode;
		}
		iCount++;
	}
	
	public void deleteFirst()
	{
		if(first == null)
		{
			System.out.println("Linked list is empty\n");		
			return;
		}
		else if(iCount == 1)
		{
			first = null;
		}
		else
		{
			first = first.nextPtr;
			first.prevPtr = null;
			iCount--;	
		}
	}
	
	public void deleteLast()
	{
		if(first == null)
		{
			System.out.println("Linked list is empty\n");		
			return;
		}
		else if(first.nextPtr == null)
		{
			first = null;
		}
		else
		{
			NodeD tempPtr = first;
			while(tempPtr.nextPtr != null)
			{
				tempPtr = tempPtr.nextPtr;
			}	
			
			tempPtr.nextPtr = null;
		}
		iCount--;
	}
	
	public void insertAtPos(int iNo,int iPos)
	{
		if((iPos < 0) || (iPos > iCount + 1))
		{
			System.out.println("Invalid position\n");
			return;
		}
		
		if(iPos == 1)
		{
			insertFirst(iNo);
		}
		else if(iPos == iCount + 1)
		{
			insertLast(iNo);
		}
		else
		{
			int iCnt = 0;
			NodeD tempPtr = first;
			
			for(iCnt = 1; iCnt < iPos - 1; iCnt++)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			NodeD newNode = null;			
			newNode = new NodeD(iNo);
			
			newNode.nextPtr = tempPtr.nextPtr;
			tempPtr.nextPtr = newNode;
			
			tempPtr.nextPtr = newNode;
			newNode.prevPtr = tempPtr;		
					
			iCount++;
		}
		
	}
	
	public void deleteAtPos(int iPos)
	{
			if((iPos < 0) || (iPos > iCount + 1))
		{
			System.out.println("Invalid position\n");
			return;
		}
		
		if(iPos == 1)
		{
			deleteFirst();
		}
		else if(iPos == iCount)
		{
			deleteLast();
		}
		else
		{
			int iCnt = 1;
			NodeD tempPtr = first;
			
			for(iCnt = 1; iCnt < iPos - 1; iCnt++)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			tempPtr.nextPtr = tempPtr.nextPtr.nextPtr;
			tempPtr.nextPtr.nextPtr = tempPtr;
			
			iCount--;
		}
		
	}
	
}

class DoublyLinearLL
{
	public static void main(java.lang.String Arg[])
	{
		DoublyLinear dObj = new DoublyLinear();
		
		dObj.insertFirst(101);
		dObj.insertFirst(51);
		dObj.insertFirst(21);
		dObj.insertFirst(11);

	System.out.println("============================================================================");
	
	
	System.out.println("\n----------------------------------------------------------------------------");
		System.out.println("Linked list after inserting nodes at first position : ");
		System.out.println("----------------------------------------------------------------------------");

		dObj.displayNodes();
		System.out.println("Number of elements in linked lists are : "+dObj.countNodes());
	
	System.out.println("\n----------------------------------------------------------------------------");	
		dObj.insertLast(111);
		dObj.insertLast(151);

		System.out.println("Linked list after inserting nodes at last position : ");
		System.out.println("----------------------------------------------------------------------------");		
		dObj.displayNodes();
		System.out.println("Number of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		dObj.insertAtPos(55,4);

		System.out.println("Linked list after inserting node at given position : ");
		System.out.println("----------------------------------------------------------------------------");		
		dObj.displayNodes();
		System.out.println("Number of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		dObj.deleteAtPos(4);

		System.out.println("Linked list after deleting node at given position : ");
		System.out.println("----------------------------------------------------------------------------");		
		dObj.displayNodes();
		System.out.println("Number of elements in linked lists are : "+dObj.countNodes());
	
	System.out.println("\n----------------------------------------------------------------------------");	
		
		dObj.deleteFirst();		
		
		System.out.println("Linked list after deleting node from first position : ");
		System.out.println("----------------------------------------------------------------------------");

		dObj.displayNodes();
		System.out.println("Number of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		
		dObj.deleteLast();
		System.out.println("Linked list after deleting node from last position : ");
		System.out.println("----------------------------------------------------------------------------");

		dObj.displayNodes();
		System.out.println("Number of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n============================================================================");

	}	
}

File: Using_Python/LinkedLists/DoublyLinearLL.py
class NodeD:
	
	def __init__(self,element):
		
		self.prevPtr = None
		self.nextPtr = None
		self.Data = element

class DoublyLinearLL:
	
	def __init__(self):	
		
		self.iCount = 0;
		self.First = None;				
	
	def CountNodes(self):
		
		return self.iCount
		
	def DisplayNodes(self):
		
		print("Elements in linked lists are : ")
		
		tempPtr = self.First
		
		for iCnt in range(0,self.iCount,1):
			print("|",tempPtr.Data,"|<=>",end = ' ')
		
		print("None")
		
	def InsertFirst(self,Value):	
		
		newNode = None
		
		newNode = NodeD(Value)
		
		if(self.iCount == 0):
			
			First = newNode
		else:
			newNode.nextPtr = self.First
			self.First = newNode
		
		self.iCount = self.iCount + 1
		
	def InsertLast(self,Value):	
		
		newNode = None
		
		newNode = NodeD(Value)
		
		if(self.iCount == 0):
			
			First = newNode
		
		else:
			tempPtr = self.First
			
			while(tempPtr.nextPtr != None):	
				tempPtr = tempPtr.nextPtr
				
			newNode.prevPtr = tempPtr
			tempPtr.nextPtr = newNode
			self.First = newNode
		
		self.iCount = self.iCount + 1
		
	
	def DeleteFirst(self):
		
		if(self.iCount == 0):
			print("Linked list is empty\n")
			return
		
		elif(self.iCount == 1):
			self.First = None
			
		else:
			self.First = self.First.nextPtr
			self.prevPtr = None
			
			self.iCount = self.iCount + 1
		
	
		def DeleteLast(self):
		
			if(self.iCount == 0):
				print("Linked list is empty\n")
				return
		
			elif(self.iCount == 1):
				self.First = None
			
			else:
		
				tempPtr = self.First
				
				while(tempPtr.nextPtr.nextPtr != None):	
					tempPtr = tempPtr.nextPtr
				
				tempPtr.nextPtr = None	
				
				self.iCount = self.iCount + 1
		
		def InsertAtPos(self,Value,Pos):

			if((Pos < 1) or (Pos > self.iCount + 1)):
				print("Invalid number of position\n")
				return

			if(Pos == 1):
				InsertFirst(Value)

			elif(Pos == self.iCount + 1):
				InsertLast(Value)

			else:
				tempPtr = self.First

				for iCnt in range(0,Pos - 1,1):

					tempPtr = tempPtr.nextPtr

				newNode = NodeD(Value)

				newNode.nextPtr = tempPtr.nextPtr
				tempPtr.nextPtr.prevPtr = newNode

				tempPtr.nextPtr = newNode
				newNode.prevPtr = tempPtr

				self.iCount = self.iCount + 1	
						

		def DeleteAtPos(Pos):
			
			if((Pos < 1) or (Pos > self.iCount)):
				print("Invalid number of position\n")
				return

			if(Pos == 1):
					DeleteFirst()

			elif(Pos == self.iCount):
					DeleteLast()

			else:
				tempPtr = self.First

				for iCnt in range(0,Pos - 1,1):

					tempPtr = tempPtr.nextPtr

				tempPtr.nextPtr = tempPtr.nextPtr.nextPtr
				tempPtr.nextPtr.prevPtr = tempPtr

				self.iCount = self.iCount + 1	
						
def main():
	
	line = "-"*50
	
	print("-"*20,"Implementation Of Doubly Linear Linked Lists","-"*20)
	
	dObj = DoublyLinearLL()
	
	dObj.InsertFirst(51)
	dObj.InsertFirst(21)
	dObj.InsertFirst(11) 	
	print("\n")
	print(line)
	print("Linked list after inserting nodes at first position")	
	print(line)
	dObj.DisplayNodes()
	print("Number of nodes in linked lists are : ",dObj.CountNodes())
	
	dObj.InsertLast(101)
	dObj.InsertLast(111)	
	print("\n")
	print(line)
	print("Linked list after inserting nodes at last position")
	print(line)
	dObj.DisplayNodes()
	print("Number of nodes in linked lists are : ",dObj.CountNodes())
	
	dObj.InsertAtPos(55,4)
	print("\n")
	print(line)
	print("Linked list after inserting node at 4 position")
	print(line)
	dObj.DisplayNodes()
	print("Number of nodes in linked lists are : ",dObj.CountNodes())
	
	dObj.DeleteAtPos(4)
	print("\n")
	print(line)
	print("Linked list after deleting node at 4 position")
	print(line)
	dObj.DisplayNodes()
	print("Number of nodes in linked lists are : ",dObj.CountNodes())

	dObj.DeleteFirst()
	print("\n")
	print(line)
	print("Linked list after deleting node from first position")
	print(line)
	dObj.DisplayNodes()
	print("Number of nodes in linked lists are : ",dObj.CountNodes())
	
	dObj.DeleteLast()
	print("\n")
	print(line)
	print("Linked list after deleting node from last position")
	print(line)
	dObj.DisplayNodes()
	print("Number of nodes in linked lists are : ",dObj.CountNodes())


if __name__ == "__main__":
	main()
